flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/ShuffleOperator.java

```diff
@@ -0,0 +1,132 @@
+  private transient DataStatistics<K> localDataStatistics;
+  private transient DataStatistics<K> globalDataStatistics;
+  private transient ListState<DataStatistics<K>> globalDataStatisticsState;
+  private transient DataStatisticsFactory<K> dataStatisticsFactory;
+
+  public ShuffleOperator(
+      KeySelector<T, K> keySelector,
+      OperatorEventGateway operatorEventGateway,
+      DataStatisticsFactory<K> dataStatisticsFactory) {
+    this.keySelector = keySelector;
+    this.operatorEventGateway = operatorEventGateway;
+    this.dataStatisticsFactory = dataStatisticsFactory;
+  }
+
+  @VisibleForTesting
+  ListStateDescriptor<DataStatistics<K>> generateGlobalDataDistributionWeightDescriptor() {
+    return new ListStateDescriptor<>(
+        "globalDataStatisticsState", TypeInformation.of(new TypeHint<DataStatistics<K>>() {}));
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    localDataStatistics = dataStatisticsFactory.createDataStatistics();
+    globalDataStatisticsState =
+        context
+            .getOperatorStateStore()
+            .getListState(generateGlobalDataDistributionWeightDescriptor());
+
+    if (globalDataStatisticsState.get() != null
+        && globalDataStatisticsState.get().iterator().hasNext()) {
+      globalDataStatistics = globalDataStatisticsState.get().iterator().next();
+    } else {
+      globalDataStatistics = dataStatisticsFactory.createDataStatistics();
+    }
+  }
+
+  @Override
+  public void open() throws Exception {
+    // TODO: handle scaling up
+    if (globalDataStatistics != null && globalDataStatistics.size() > 0) {
+      output.collect(new StreamRecord<>(ShuffleRecordWrapper.fromStatistics(globalDataStatistics)));
+    }
+  }
+
+  @Override
+  public void handleOperatorEvent(OperatorEvent evt) {
+    // TODO: receive event with globalDataDistributionWeight from coordinator and update
+    // globalDataStatistics
+  }
+
+  @Override
+  public void processElement(StreamRecord<T> streamRecord) throws Exception {
+    final K key = keySelector.getKey(streamRecord.getValue());
+    localDataStatistics.put(key);
+    output.collect(new StreamRecord<>(ShuffleRecordWrapper.fromRecord(streamRecord.getValue())));
+  }
+
+  @Override
+  public void snapshotState(StateSnapshotContext context) throws Exception {
+    long checkpointId = context.getCheckpointId();
+    LOG.debug("Taking shuffle operator snapshot for checkpoint {}", checkpointId);
+
+    // Set the globalDataDistributionWeightState state to the latest global value.
+    if (globalDataStatistics != null && globalDataStatistics.size() > 0) {
+      globalDataStatisticsState.clear();
+      globalDataStatisticsState.add(globalDataStatistics);
+    }
+
+    // TODO: send to coordinator
+    // For now we make it simple to send globalDataStatisticsState at checkpoint
+
+    // Reset the local data count
+    localDataStatistics = dataStatisticsFactory.createDataStatistics();
```

Comment:
It might make sense to add a `reset` or `clear` method to `DataStatistics` interface. what do you think? @yegangy0718 and @pvary

Reply:
Either `reset` or `clear` method - or a factory for the `DataStatistics`.
TDigest does not have a `reset` or `clear` like method, but we can hide it if we want to use it later. I think we should decide what creates a cleaner API.